src/software/gui/robot_diagnostics/widgets/robot_diagnostics_gui.cpp

```diff
@@ -33,24 +31,155 @@ void RobotDiagnosticsGUI::onChickerStateChanged(double chicker_power,
     double direct_per_wheel_power, DirectPerWheelMode direct_per_wheel_mode)
 {
-    // TODO (Issue #1420): push primitive set msg to buffer
+    auto direct_control_primitive = createDirectControlPrimitiveFromUI();
+    sendPrimitiveSet(std::move(direct_control_primitive));
 }
 
 void RobotDiagnosticsGUI::onDribblerPowerChanged(double dribbler_power)
 {
-    // TODO (Issue #1420): push primitive set to buffer
+    auto direct_control_primitive = createDirectControlPrimitiveFromUI();
+    sendPrimitiveSet(std::move(direct_control_primitive));
+}
+
+std::unique_ptr<TbotsProto::Primitive>
+RobotDiagnosticsGUI::createDirectControlPrimitiveFromUI()
+{
+    auto direct_control_primitive_msg = std::make_unique<TbotsProto::Primitive>();
+
+    switch (main_widget->tabWidget->currentIndex())
+    {
+        case WheelControlTab::DIRECT_PER_WHEEL:
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_front_left_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_fl->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_front_right_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_fr->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_back_left_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_bl->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_per_wheel_control()
+                ->set_back_right_wheel_rpm(
+                    main_widget->lineEdit_direct_per_wheel_br->text().toFloat());
+            break;
+        case WheelControlTab::DIRECT_VELOCITY:
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_velocity_control()
+                ->mutable_velocity()
+                ->set_x_component_meters(
+                    main_widget->lineEdit_direct_velocity_x->text().toFloat());
+            direct_control_primitive_msg->mutable_direct_control()
+                ->mutable_direct_velocity_control()
+                ->mutable_velocity()
+                ->set_y_component_meters(
+                    main_widget->lineEdit_direct_velocity_y->text().toFloat());
+
+            auto angular_velocity_msg =
+                createAngularVelocityProto(AngularVelocity::fromDegrees(
+                    main_widget->lineEdit_direct_velocity_theta->text().toDouble()));
+            *(direct_control_primitive_msg->mutable_direct_control()
+                  ->mutable_direct_velocity_control()
+                  ->mutable_angular_velocity()) = *angular_velocity_msg;
+            break;
+    }
+
+    if (main_widget->buttonGroup_charge_state->checkedButton() ==
```

Comment:
Similarly, some comments can be helpful here.

Reply:
added